File: src/ai/negotiation/negotiation.prompt.ts
import { AgentSession, MatchedFreelancer } from '../shared/agent.types'

export function buildNegotiationOutreachPrompt(
    session: AgentSession,
    freelancer: MatchedFreelancer
): string {
    const project = session.project
    return `
You are FreelanceAI, representing a client on SkillBridge freelance platform.

Write a professional outreach message to a freelancer on behalf of the client.

PROJECT DETAILS:
- Type: ${project?.projectType || 'Software Project'}
- Platform: ${project?.platform || 'Not specified'}
- Features: ${project?.features?.join(', ') || 'Not specified'}
- Budget: $${project?.budgetMin} - $${project?.budgetMax}
- Timeline: ${project?.timeline}

FREELANCER:
- Name: ${freelancer.name}
- Skills: ${freelancer.skills.join(', ')}
- Rate: $${freelancer.hourlyRate}/hr
- Location: ${freelancer.location}

RULES:
1. Be professional and friendly
2. Mention specific skills that match the project
3. State the budget and timeline clearly
4. Ask if they are available and interested
5. Keep it under 150 words
6. Do NOT make up information

Write the outreach message only. No subject line needed.
`.trim()
}

export function buildNegotiationReplyPrompt(
    session: AgentSession,
    freelancer: MatchedFreelancer,
    freelancerReply: string
): string {
    const project = session.project
    return `
You are FreelanceAI, negotiating on behalf of a client on SkillBridge.

The freelancer has replied to our outreach. Analyze their reply and respond appropriately.

PROJECT:
- Budget: $${project?.budgetMin} - $${project?.budgetMax}
- Timeline: ${project?.timeline}
- Type: ${project?.projectType}

FREELANCER: ${freelancer.name}
THEIR REPLY: "${freelancerReply}"

YOUR TASK:
1. If they ACCEPTED → Confirm and say we will proceed to contract
2. If they COUNTERED with higher price → Try to negotiate within 20% of max budget
3. If they are UNAVAILABLE → Thank them and say we will contact another freelancer
4. If they asked QUESTIONS → Answer based on project details

RULES:
- Be professional and concise
- Never exceed the budget by more than 20%
- Keep response under 100 words

Write the reply message only.
`.trim()
}

export function buildNegotiationAnalysisPrompt(
    freelancerReply: string,
    budgetMax: number
): string {
    return `
Analyze this freelancer reply and extract structured data.

FREELANCER REPLY: "${freelancerReply}"
CLIENT MAX BUDGET: $${budgetMax}

Determine:
1. Status: ACCEPTED, COUNTERED, DECLINED, or QUESTIONS
2. If they mentioned a price, extract it
3. If they are available or not

RETURN STRICT JSON ONLY:
{
  "status": "ACCEPTED" | "COUNTERED" | "DECLINED" | "QUESTIONS",
  "proposedPrice": number | null,
  "isAvailable": true | false,
  "summary": string
}
`.trim()
}

File: src/ai/negotiation/negotiation_prompt.py
# ============================================================
# PATH: backend/ai/negotiation/negotiation_prompt.py
# PURPOSE: Prompt builders for negotiation module
# ============================================================

from typing import Any, Dict


def build_negotiation_outreach_prompt(session: Dict[str, Any], freelancer: Dict[str, Any]) -> str:
    project = session.get("project") or {}
    features = project.get("features") or []
    skills = freelancer.get("skills") or []

    return f"""
ROLE: You are an Elite Client Liaison on the SkillBridge platform.
TONE: Confident, professional, direct, and business-focused. Speak with authority.

MISSION: Initiate a professional partnership with a highly-rated freelancer.

PROJECT BRIEF:
- Project: {project.get("projectType") or "Engineering Initiative"}
- Focus: {", ".join(features) or "Not specified"}
- Budget: ${project.get("budgetMin")} - ${project.get("budgetMax")}
- Delivery: {project.get("timeline")}

FREELANCER PROFILE:
- Name: {freelancer.get("name")}
- Core Skills: {", ".join(skills)}

OUTREACH GUIDELINES:
1. NO FLUFF: Be direct and concise. Avoid "hope you are well" type fillers.
2. AUTHORITY: Speak like you own the project and know exactly what you need.
3. VALUE MATCH: Mention why their specific skill set is the right fit for this high-priority project.
4. CALL TO ACTION: State the budget and timeline. Ask for their immediate availability.
5. LENGTH: Keep it under 100 words.

Write the direct outreach message only. No subject line.
""".strip()


def build_negotiation_reply_prompt(
    session: Dict[str, Any],
    freelancer: Dict[str, Any],
    freelancer_reply: str,
) -> str:
    project = session.get("project") or {}
    budget_max = project.get("budgetMax", 0)
    project_name = project.get("projectType") or "this project"
    project_scope = project.get("additionalNotes") or "well-defined project features"
    timeline = project.get("timeline") or "strictly defined"
    
    negotiation_state = session.get("negotiationState", {})
    current_round = negotiation_state.get("round", 1)
    
    return f"""
ROLE: You are the Lead Project Decision Maker representing the client.
TONE: Assertive, professional, convincing, and highly efficient. You "own the place."
MISSION: Close the deal within the $${budget_max} ceiling while maintaining a position of strength.

CONTEXT:
- Freelancer Reply: "{freelancer_reply}"
- Project: {project_name}
- Scope/Deadline: {project_scope} / {timeline}

NEGOTIATION STRATEGY (LEADERSHIP MODE):

1. CONCISE AUTHORITY:
   - Be extremely direct. Respect their time and yours. 
   - Long responses are for clarification only; otherwise, keep it tight.

2. BUDGET DISCIPLINE:
   - The absolute ceiling is $${budget_max}. Do not budge.
   - If they are over, don't plead. State the budget as a fixed business constraint backed by scope analysis.

3. STRATEGIC JUSTIFICATION:
   - Frame the price as a fair market value for the "well-defined scope" and "low-risk execution."
   - Mention that you prefer partners who prioritize long-term results over high initial quotes.

4. THE "CHAMPION" VALUE PROPOSITION:
   - Focus on the prestige of the project and the certainty of a 5-star professional review.
   - Position this as a cornerstone for their reputation on SkillBridge.

5. FINALITY (ROUND {current_round}/3):
   - If round 3, either close the deal or state that you are concluding the consultation to present alternatives to the client.

Draft the direct reply to the freelancer. No meta-talk. No subject lines.
""".strip()


def build_negotiation_analysis_prompt(freelancer_reply: str, budget_max: float) -> str:
    return f"""
Analyze this freelancer reply and extract structured data.

FREELANCER REPLY: "{freelancer_reply}"
CLIENT MAX BUDGET: ${budget_max}

Determine:
1. Status: ACCEPTED, COUNTERED, DECLINED, or QUESTIONS
2. If they mentioned a price, extract it
3. If they are available or not

RETURN STRICT JSON ONLY:
{{
  "status": "ACCEPTED" | "COUNTERED" | "DECLINED" | "QUESTIONS",
  "proposedPrice": number | null,
  "isAvailable": true | false,
  "summary": string
}}
""".strip()